### PR TITLE
Clicking on layer example

### PR DESCRIPTION
**Pull request by langsmith**
This PR adds an example of clicking on a fillLayer. Almost all of the code comes from @tobrun in [this ticket](https://github.com/mapbox/mapbox-android-demo/issues/305).

cc @Guardiola31337 

device-2017-07-08-211512

### CODE REVIEW

File: MapboxAndroidDemo/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -75,6 +75,13 @@
                 android:value="com.mapbox.mapboxandroiddemo.MainActivity"/>
         </activity>
         <activity
+            android:name=".examples.annotations.ClickOnLayerActivity"
+            android:label="@string/activity_annotation_geojson_line_title">
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/AndroidManifest.xml]: `activity_annotation_click_on_layer_title`

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
+package com.mapbox.mapboxandroiddemo.examples.annotations;
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: This seems more like a query map or runtime styling example rather than annotation.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+
+import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.fillOpacity;
+
+/**
+ * Detect click events on a polygon that was added as a GeoJsonSource.
+ */
+
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: nit extra line

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+public class ClickOnLayerActivity extends AppCompatActivity {
+
+  private MapView mapView;
+  private String geoJsonSourceId = "GeoJSON source";
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: Can be static String using a more descriptive, single word, ID

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  private String geoJsonLayerId = "GeoJSON layer";
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: Can be static String using a more descriptive, single word, ID

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+          // Load GeoJSONSource
+          source = new GeoJsonSource(geoJsonSourceId, new URL("https://gist.githubusercontent"
+              + ".com/tobrun/cf0d689c8187d42ebe62757f6d0cf137/raw/4d8ac3c8333f1517df9d303"
+              + "d58f20f4a1d8841e8/regions.geojson"));
+
+          // Add GeoJsonSource to map
+          mapboxMap.addSource(source);
+
+        } catch (MalformedURLException malformedUrlException) {
+          Log.d("Invalid URL", "malformedUrlException = " + malformedUrlException.toString());
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: Log.e and use `:` rather than `=`

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        Toast.LENGTH_SHORT).show();
+
+    mapView = (MapView) findViewById(R.id.mapView);
+    mapView.onCreate(savedInstanceState);
+    mapView.getMapAsync(new OnMapReadyCallback() {
+      @Override
+      public void onMapReady(final MapboxMap mapboxMap) {
+
+        try {
+
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: nit extra line

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    Toast.makeText(ClickOnLayerActivity.this, R.string.click_on_polygon_toast_instruction,
+        Toast.LENGTH_SHORT).show();
+
+    mapView = (MapView) findViewById(R.id.mapView);
+    mapView.onCreate(savedInstanceState);
+    mapView.getMapAsync(new OnMapReadyCallback() {
+      @Override
+      public void onMapReady(final MapboxMap mapboxMap) {
+
+        try {
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: Might be better to move this entire try/catch block to a private method.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            for (com.mapbox.services.commons.geojson.Feature feature : featureList) {
+              Log.d("Feature found with %1$s", feature.toJson());
+
+              Toast.makeText(ClickOnLayerActivity.this, R.string.click_on_polygon_toast,
+                  Toast.LENGTH_SHORT).show();
+
+            }
+          }
+        });
+
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: nit extra line

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+            RectF rectF = new RectF(pointf.x - 10, pointf.y - 10, pointf.x + 10, pointf.y + 10);
+
+            List<Feature> featureList = mapboxMap.queryRenderedFeatures(rectF, geoJsonLayerId);
+
+            for (com.mapbox.services.commons.geojson.Feature feature : featureList) {
+              Log.d("Feature found with %1$s", feature.toJson());
+
+              Toast.makeText(ClickOnLayerActivity.this, R.string.click_on_polygon_toast,
+                  Toast.LENGTH_SHORT).show();
+
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: nit extra line

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  @Override
+  protected void onSaveInstanceState(Bundle outState) {
+    super.onSaveInstanceState(outState);
+    mapView.onSaveInstanceState(outState);
+  }
+
+  public boolean onOptionsItemSelected(MenuItem item) {
+    switch (item.getItemId()) {
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: Back button should be handled in AppCompat without the need for this activity method.

File: MapboxAndroidDemo/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -140,6 +141,7 @@
     <string name="activity_annotation_marker_view_description">Attach a view to a given position on the map.</string>
     <string name="activity_annotation_custom_info_window_description">Use an info window adapter to customize the info window.</string>
     <string name="activity_annotation_animated_marker_description">Animate the marker to a new position on the map.</string>
+    <string name="activity_annotation_click_on_layer_description">Click on and highlight a selected GeoJSON polygon</string>
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/res/values/strings.xml]: missing period

File: MapboxAndroidDemo/src/main/AndroidManifest.xml (hunk 1 above)
Comment (tobrun) [inline MapboxAndroidDemo/src/main/AndroidManifest.xml]: this example is unrelated to the term "annotation", I would scratch that.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 4 above)
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: and final :)

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 5 above)
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: and final :)

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 6 above)
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: don't see a technical reason to prefer `=` over `:`.

When logging exceptions you should use the 3 args Log methods:
```
Log.e(String tag, String msg, Throwable tr) 
```
No need for concatenation, overall a bit cleaner imo

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 11 above)
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: the term AppCompat refers to the configuration you can do in `AndroidManifest.xml`, where you can provide a relation between the child and parent activity.

**Comment by tobrun** [comment]
Before merging we need to look into if we need to attribute the source of the data used. 
The geojson file is coming from https://mapzen.com/data/borders/.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 6 above)
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: 👍  agreed, been using Timber too much

File: MapboxAndroidDemo/src/main/AndroidManifest.xml (hunk 1 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/AndroidManifest.xml]: changed to `query` section

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 2 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: moved to `query` section

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 3 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 7 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 10 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 9 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: removed

File: MapboxAndroidDemo/src/main/res/values/strings.xml (hunk 12 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/res/values/strings.xml]: period added

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 5 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 4 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 8 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: done

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 6 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: Fixed

**Comment by langsmith** [comment]
@tobrun , regarding your attribution comment above. I emailed MapZen and they said 

screen shot 2017-07-19 at 2 33 26 pm

@JustinVE , the context here is that we're using a GeoJSON file of the Belgium border in this example. Do you know whether our own map attribution covers us for using this OSM data? Or do we need to add some sort of other attribution?

**Comment by langsmith** [comment]
@cammace , how do you propose combining this example with https://github.com/mapbox/mapbox-android-demo/blob/master/MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/SelectBuildingActivity.java? 

Yes, they both `queryRenderedFeatures()` and other similar code,  but they seem like two good and distinct enough use cases for keeping them separate 😕

**Comment by JustinVE** [comment]
@langsmith - I assume this is on a Mapbox map, and using our standard attribution?  In that case - yes, our standard OSM attribution will cover this use of OSM data.

**Comment by langsmith** [comment]
Ok, thanks @JustinVE . Yep, it's on a Mapbox map with our standard attribution

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java (hunk 11 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/annotations/ClickOnLayerActivity.java]: Have added `android:launchMode="singleTop"` to manifest file to replace `onOptionsItemSelected`

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/ClickOnLayerActivity.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    mapView.onDestroy();
+  }
+
+  @Override
+  protected void onSaveInstanceState(Bundle outState) {
+    super.onSaveInstanceState(outState);
+    mapView.onSaveInstanceState(outState);
+  }
+
+  public boolean onOptionsItemSelected(MenuItem item) {
```
Comment (cammace) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/ClickOnLayerActivity.java]: We can remove this now correct?

Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/ClickOnLayerActivity.java]: removed

**Comment by caopeng000** [comment]
I want to implement nested geo fences.How to do it？
ymgxc bbv s4hwd 7y ybu
I am an android development engineer

**Comment by langsmith** [comment]
@caopeng000 ,

Please see this example in the demo app. You:

1. [Create a source](https://github.com/mapbox/mapbox-android-demo/blob/bc613d62da613d964b4a7a5d8e5018a7f274047b/MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/ClickOnLayerActivity.java#L87-L100)
2. [Add the source to the map](https://github.com/mapbox/mapbox-android-demo/blob/bc613d62da613d964b4a7a5d8e5018a7f274047b/MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/ClickOnLayerActivity.java#L62-L65)
3. [Detect whether the click was inside the polygon or not](https://github.com/mapbox/mapbox-android-demo/blob/bc613d62da613d964b4a7a5d8e5018a7f274047b/MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/ClickOnLayerActivity.java#L67-L82)
